File: cpp/oop/basic_class.cpp
#include <iostream>
#include <string>

class BankAccount{
    private:
        std::string account_num;
        int balance;

    public:
        BankAccount(std::string account_num, int balance) : account_num(account_num), balance(balance){}
    
        float get_account_balance(){
            return balance;
        }

        std::string get_account_num(){
            return account_num;
        }

        void deposit(float amount){
            balance += amount;
        }

        void withdraw(float amount){
            if (balance >= amount){
                balance -= amount;
            } else {
                std::cout<<"Insufficient funds!"<<std::endl;
            }
        }

};

int main(){

    BankAccount myaccount("1234 5678 90", 1000);
    std::cout<<myaccount.get_account_num()<<std::endl;          // 1234 5678 90
    std::cout<<myaccount.get_account_balance()<<std::endl;      // 1000
    myaccount.deposit(1000);
    myaccount.withdraw(100);
    myaccount.withdraw(2000);                                   // Insufficient funds!
    return 0;
}

File: python/oop/basic_class.py
"""
Python has no private variables like in cpp.
You can use name conventions for private variables in python. 
Name mangling is used here.
"""

class Bankaccount():
    def __init__(self, account_num : str, balance : float):
        self.__account_num = account_num
        self.__balance = balance
    
    def get_account_balance(self) -> float:
        return self.__balance
    
    def get_account_number(self) -> str:
        return self.__account_num

    def deposit(self, amount : float) -> None:
        self.__balance += amount
    
    def withdraw(self, amount : float) -> None:
        if self.__balance >= amount:
            self.__balance -= amount
        else:
            print("Insufficient Funds!")

account = Bankaccount("1234 5678 90", 1000)
#print(account.__account_num)           # This will give error. 
#print(account.__balance)               # 'Bankaccount' object has no attribute '__balance'

print(account.get_account_number())     # 1234 5678 90
print(account.get_account_balance())    # 1000

account.deposit(1000)
account.withdraw(100)
account.withdraw(2000)                  # Insufficient funds!

